src/Data/List/Ext/Subperm.agda

```diff
@@ -0,0 +1,37 @@
+{-# OPTIONS --safe #-}
+
+module Data.List.Ext.Subperm where
+
+open import Prelude
+
+open import Data.List
+open import Data.List.Relation.Unary.Any using (here; there; any?)
+open import Relation.Binary.Definitions using (Decidable)
+open import Data.List.Membership.Propositional using (_∈_)
+
+module _ {A : Set} where
+
+  open import Relation.Binary.Core using (REL)
+
+  data Subperm : REL (List A) (List A) 0ℓ where
+    []   : Subperm [] []
+    _∷ₚʳ_ : ∀ {xs ys} → ∀ y → Subperm xs ys → Subperm xs (y ∷ ys)
+    _∷ₚ_  : ∀ {x xs y ys} → x ∈ (y ∷ ys) → Subperm xs (y ∷ ys) → Subperm (x ∷ xs) (y ∷ ys)
+
+  Subperm⁻¹ : {x : A}{xs l : List A} → Subperm (x ∷ xs) l → x ∈ l × Subperm xs l
+  Subperm⁻¹ (y ∷ₚʳ Spxxs) = (there (proj₁ (Subperm⁻¹ Spxxs))) , (y ∷ₚʳ proj₂ (Subperm⁻¹ Spxxs))
+  Subperm⁻¹ (x∈yys ∷ₚ Spxsyys) = x∈yys , Spxsyys
+
+  Subperm[] : {l : List A} → Subperm [] l
+  Subperm[] {[]} = []
+  Subperm[] {x ∷ _} = x ∷ₚʳ Subperm[]
```

Comment:
1. Note that the first argument of `_∷ₚʳ_` (i.e. the element being "dropped" from the sub-permutation) can (almost) always be inferred from the second type-level index of the `Subperm` relation (i.e. the big list). So you will (almost) never need to explicitly give it in terms, which suggests a "cons" infix notation might not be the best option here.

2. Another confusing aspect of using two "cons" operators for the two (recursive) constructors is that the element being prepended is different in each case: for `_∷ₚʳ_` we are "dropping" a single element; for `_∷ₚ_` we are "keeping" a proof of membership (of a single element).

3. Constructor overloading is one of the biggest conveniences we have in Agda, so I would not be afraid to use the more idiomatic (in the context of stdlib) `keep` and `drop` for the (recursive) constructors instead. Coupled with points (1) and (2), here's the definition I would write to express the same thing:
```agda
  data Subperm : Rel (List A) 0ℓ where
    []   : Subperm [] []
    drop : ∀ {xs ys y} → Subperm xs ys → Subperm xs (y ∷ ys)
    keep : ∀ {x xs ys} → x ∈ ys → Subperm xs ys → Subperm (x ∷ xs) ys
```
The resulting definitions you get become simpler and more intuitive IMO:
```agda
  Subperm⁻¹ : {x : A}{xs l : List A} → Subperm (x ∷ xs) l → x ∈ l × Subperm xs l
  Subperm⁻¹ (drop p)    = bimap there drop (Subperm⁻¹ p)
  Subperm⁻¹ (keep x∈ p) = x∈ , p

  Subperm[] : {l : List A} → Subperm [] l
  Subperm[] {[]} = []
  Subperm[] {_ ∷ _} = drop Subperm[]

  subperm? : ⦃ DecEq A ⦄ → Decidable (Subperm)
  subperm? [] _ = yes Subperm[]
  subperm? (_ ∷ _) [] = no λ where (keep () _)
  subperm? (x ∷ xs) (y ∷ ys)
    with ¿ x ∈ (y ∷ ys) ¿
  ... | no  x∉ = no (x∉ ∘ proj₁ ∘ Subperm⁻¹)
  ... | yes x∈
    with subperm? xs (y ∷ ys)
  ... | no ¬p = no (¬p ∘ proj₂ ∘ Subperm⁻¹)
  ... | yes p = yes (keep x∈ p)
```